_posts/2022-10-06-audio-classification-torch/audio-classification.Rmd

```diff
@@ -0,0 +1,472 @@
+                        n_fft = 512,
+                        window_size_seconds = 0.03,
+                        window_stride_seconds = 0.01,
+                        power = 2) {
+    self$pad_to <- pad_to
+    self$window_size_samples <- sampling_rate *
+      window_size_seconds
+    self$window_stride_samples <- sampling_rate *
+      window_stride_seconds
+    self$power <- power
+    self$spectrogram <- transform_spectrogram(
+        n_fft = n_fft,
+        win_length = self$window_size_samples,
+        hop_length = self$window_stride_samples,
+        normalized = TRUE,
+        power = self$power
+      )
+    super$initialize(...)
+  },
+  .getitem = function(i) {
+    item <- super$.getitem(i)
+
+    x <- item$waveform
+    # make sure all samples have the same length (57)
+    # shorter ones will be padded,
+    # longer ones will be truncated
+    x <- nnf_pad(x, pad = c(0, self$pad_to - dim(x)[2]))
+    x <- x %>% self$spectrogram()
+
+    if (is.null(self$power)) {
+      # in this case, there is an additional dimension, in position 4,
+      # that we want to appear in front
+      # (as a second channel)
+      x <- x$squeeze()$permute(c(3, 1, 2))
+    }
+
+    y <- item$label_index
+    list(x = x, y = y)
+  }
+)
+
+```
+
+In the parameter list to `spectrogram_dataset()`, note `power`, with a default value of 2. This is the value that, unless told otherwise, `torch`'s `transform_spectrogram()` will assume that `power` should have. Under these circumstances, the values that make up the spectrogram are the squared magnitudes of the Fourier coefficients. Using `power`, you can change the default, and specify, for example, that'd you'd like absolute values (`power = 1`), any other positive value (such as `0.5`, the one we used above to display a concrete example) -- or both the real and imaginary parts of the coefficients (`power = NULL)`.
+
+Display-wise, of course, the full complex representation is inconvenient; the spectrogram plot would need an additional dimension. But we may well wonder whether a neural network could profit from the additional information contained in the "whole" complex number. After all, when reducing to magnitudes we lose the phase shifts for the individual coefficients, which might contain usable information. In fact, my tests showed that it did; use of the complex values resulted in enhanced classification accuracy. [^1]
+
+[^1]: In the book, I'm comparing three different setups (the third being use of Mel spectrograms). Using complex coefficients for network input was the winner.
+
+Let's see what we get from `spectrogram_dataset()`:
+
+```{r}
+ds <- spectrogram_dataset(
+  root = "~/.torch-datasets",
+  url = "speech_commands_v0.01",
+  download = TRUE,
+  power = NULL
+)
+
+dim(ds[1]$x)
+```
+
+    [1]   2 257 101
+
+We have two hundred fifty-seven coefficients for a hundred and one windows; and each coefficient is represented by both its real and imaginary parts.
```

Comment:
```suggestion
We have 257 coefficients for 101 windows; and each coefficient is represented by both its real and imaginary parts.
```

Reply:
Numbers over ten (or twenty, depending on whom you ask) are written in Arabic numerals. ✅

Reply:
oh, see, so I over-generalized from some correction you made, in the past :-)